sympy/solvers/ode/ode.py

```diff
@@ -1167,8 +1162,9 @@ class in it.  Note that a hint may do this anyway if
             roots = solve(reduced_eq, df)
             if roots:
                 meq = Mul(*[(df - i) for i in roots])*Dummy()
-                m = _ode_factorable_match(meq, func, kwargs.get('x0', 0))
-                matching_hints['factorable'] = m
+                m = SingleODEProblem(meq, func, x, prep=False)
+                matching_hints['factorable'] = Factorable(m)
```

Comment:
Why isn't this calling Factorable.match?

Reply:
This logic should be in Factorable.match

Reply:
beacuse it will return true only,there is no chance of returning False as the equation passed is already given as product of its factors.

Reply:
> This logic should be in Factorable.match

@oscarbenjamin this code block was added in #15889 as `_ode_factorable_match` was not able to solve for `df`. so should i move this in `_ode_factorable_match`

Reply:
It should be in Factorable.match. At the end of this refactor all of the code from `classify_ode` should be moved into the SingleODESolver classes.

Reply:
@oscarbenjamin when i move this block into Factorable.match then lot of exceptions ar occuring for recurrsion depth

Reply:
The code for `Factorable._matches` should be simplified. The code above does not literally need to be included there. Rather the fact that it attempts solve should be. Probably `Factorable._matches` should use `factor_list` rather than checking for Pow and calling factor. Maybe if it did that then this call to solve would not be needed.

Reply:
In fact maybe `solve` is all that's needed and `Factorable._matches` can just be
```
self.solutions = solve(eq, df)
return len(self.solutions) > 1
```

Reply:
> In fact maybe `solve` is all that's needed and `Factorable._matches` can just be
> 
> ```
> self.solutions = solve(eq, df)
> return len(self.solutions) > 1
> ```
@oscarbenjamin as `solve(eq,df)` returns set of solution so we might miss many cases like
`eq = f(x).diff(x)**2 + 2*f(x).diff(x) + 1`

Reply:
Good point. I guess it could be:
```
if len(solutions) > 1:
    return True
elif len(solutions) == 1:
    return df - solutions[0] != eq # Avoid infinite recursion
```
When we're done refactoring things out of `classify_ode` we should get rid of the factorable hint and just make it so that trying to solve for df is one of the first things that happens in `dsolve`.

Probably we do still need to try calling factor as well but should use `factor_list`.

Reply:
@oscarbenjamin  if we try to use `factor_list` then the equation should always be polynomial so 2 cases should be taken care of :

    type1: -f(x) - exp(-1)*exp(x) + Derivative(f(x), (x, 2))
    type2: Derivative(f(x), x) + (2*f(x) + 3)/(x + 3)
 type2 can be manipulated easily by applying `fraction(factor(type2_eq))`